### ISSUE TITLE
CWA QR-Code certificates not scannable with check-apps

### ISSUE THREAD

**Issue by merlinseeling** [corona-warn-app/cwa-app-ios#3515]
## Avoid duplicates

* [x] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [X] Bug is specific for iOS only, for general issues / questions that apply to iOS and Android please raise them in the [documentation repository](https://github.com/corona-warn-app/cwa-documentation)
* [X] Bug is not already reported in another issue

## Technical details

- Device name: iPhone 12 and other iOS devices
- iOS version: 14.7+ & 15.X
- App version: 2.9.X

## Describe the bug

The certificate (QR-Code) that is shown in the corona-warn-app (CWA) on iOS isn’t scannable with covid-certificate-check apps like the covpass-check-app (CpCA) on android devices.

## Steps to reproduce the issue

Try to scan the QR-Code in the CWA on iOS with the CpCA from an android device. And nothing should happen. No error. Nothing. The check-app on android isn’t able to scan it at all.

## Expected behaviour

The check-app should be able to scan it and should either say that the certificate is valid or not.

## Possible Fix

After a lot of investigation I’m pretty sure I found the problem, but I don’t know how to fix it. 

The CWA is not showing the QR-Code correctly:

CWA:

<img width="600" alt="CpA" src="https://user-images.githubusercontent.com/76479756/132982911-e18393d7-c99a-4f65-914b-8cbabdccab7a.png">

covpass-app (CpA) for reference:

<img width="600" alt="CWA" src="https://user-images.githubusercontent.com/76479756/132982903-3fc7a0ec-9644-42c7-b519-7352e91a5676.png">

The Code should be made of equally sized squares, but instead the quadrangles are all differently shaped so that the QR-Code can’t be scanned by android devices.

## Additional context

I didn’t see the bug in the CpA or other apps from other countries.

There is also no problem in these constellations:

* iOS with CWA - iOS with CpCA
* android CWA - iOS with CpCA
* android CWA - android with CpCA

(CpCA is exemplary for all check-apps)

## Contact 

By any questions belonging the bug I am pleased to help, because I guess the problem is pretty important and should be fixed asap.

**Comment by Ein-Tim** [comment]
I **cannot** reproduce this bug.

Device showing the certificate:
iPhone XR with CWA version 2.9.0 and iOS 14.7.1.

Device scanning the certificate:
Huawei P Smart 2019 with CovPassCheck version 1.95.12 and Android 10.

Which device is CovPassCheck running on @merlinseeling?

**Comment by thomasaugsten** [comment]
The recommended scanning distance is 50cm. Can you please retry this or try a different android device.

**Comment by Ein-Tim** [comment]
@merlinseeling 

Thanks for providing the additional informations! Where did you download CovPassCheck from? The PlayStore?

> Another question, can you (@Ein-Tim and @thomasaugsten) see the same incorrect appearance of the QR-Code in the CWA, that I described in "Possible Fix #3515"?

Here is a part of my vaccination certificate:

Corona-Warn-App:
![IMG_1773](https://user-images.githubusercontent.com/67682506/132986200-9e6124ad-9307-4649-8863-46573ef299a7.jpg)

CovPass-App:
![IMG_1774](https://user-images.githubusercontent.com/67682506/132986214-f4a6252d-3fa4-4bbf-8927-0a8ae7437445.jpg)

**Comment by thomasaugsten** [comment]
@merlinseeling  This is not reproducable thats why we need more information how you scan.  You can also provide a video to thomas.augsten@sap.com
The thickness of the qr-codes line should not influence the correctness of the qr code

**Comment by merlinseeling** [comment]
Thats strange, by 5 iOS devices that I saw, the QR-Code was not displayed correctly. @Ein-Tim is the screenshot from the CWA from an iOS device?

>@Ein-Tim
>Where did you download CovPassCheck from? The PlayStore?

Yes!

>@thomasaugsten
>The thickness of the qr-codes line should not influence the correctness of the qr code

For my testing it does. Because it works if the QR-Code thickness is correct.

>@thomasaugsten
>This is not reproducable thats why we need more information how you scan. You can also provide a video to thomas.augsten@sap.com

Yes

**Comment by ndegendogo** [comment]
@merlinseeling could you please specify in more detail **how** you show the certificate?
- you press the 'Certificate' tab - right?
- the overview of certificates appears
- do you then show it directly from this overview, or do you select it / tap on the QR code?
- if you select it, it shows a list of certificates for a specific person.
- do you show it from this list?
- if you scroll down and select a specific certificate from this list, it shows all details for this certificate.
- do you show it from the details view?

**Comment by merlinseeling** [comment]
> * you press the 'Certificate' tab - right?

Yes

> * the overview of certificates appears

Yes

> * do you then show it directly from this overview, or do you select it / tap on the QR code?

It does not work in both cases

> * if you select it, it shows a list of certificates for a specific person.

Yes

> * do you show it from this list?

Yes 

> * if you scroll down and select a specific certificate from this list, it shows all details for this certificate.

Yes 

> * do you show it from the details view?

Yes

**Comment by ndegendogo** [comment]
> 5 iOS devices 

@merlinseeling could you please give more details on the device types?
*(in your OP you mentioned iPhone 12 - what were the others?)*

**Comment by Ein-Tim** [comment]
@merlinseeling 

> Thats strange, by 5 iOS devices that I saw, the QR-Code was not displayed correctly. @Ein-Tim is the screenshot from the CWA from an iOS device?

Yes

**Comment by thomasaugsten** [comment]
@merlinseeling you can also try to make a screenshot show it on a monitor display and try to scan it to prove the we code is correct

**Comment by merlinseeling** [comment]
> @merlinseeling could you please give more details on the device types?
> _(in your OP you mentioned iPhone 12 - what were the others?)_

Yes as mentioned before, there where may, for example 2 iPhone SE 2020, an iPhone 12 and an iPhone 12 mini

**Comment by merlinseeling** [comment]
Guys is it possible that we meet at a Video Call?
I guess that's the easiest way for me, to show the Problem to you

**Comment by thomasaugsten** [comment]
I don't think this is necessary please try to scan a iOS screenshot from a of monitor display. And if you have the time please record a video.

**Comment by merlinseeling** [comment]
I tried it again and after 50 times it worked. I don't know how and why, but it worked, but only two times and only after holding the android device still for more than 30 seconds. I didn't managed to record it, because the CpCA is restricted so that I can not record the screen while it is open. 

The problem now is only that it takes too long and the conditions must be perfect (both devices must be perfectly still and the lighting must be ideal), not that it doesn't work at all. If I try to scan with the same android device the CpA Code it works with in seconds. 

I think, that the if the QR-Code would be shown correctly, there wouldn't be a problem. I don't know why it is shown in this bad way and I tried importing the certificate again, but it wouldn't fix the incorrect appearance of the QR-Code.

The purpose of those certificates (QR-Codes) is that it is easily checkable if someone is vaccinated, recovered or tested, but if at an entrance to a restaurant or at the airport the checking takes over 2 minutes and if it isn't even clear if it works it's so annoying that I have to download the CpA because The CWA isn't even able to display a QR code correctly.

I hope I don't seem too aggressive, but I hope you understand what I'm getting at.

**Comment by Ein-Tim** [comment]
To be honest, this seems like a problem with CovPassCheck rather than with CWA. 

The QR-Code is alright how it is, I could also create a QR-Code like this and my QR-Code scanner is able to scan it: 
![frame](https://user-images.githubusercontent.com/67682506/132993303-9eb0bef8-f1c1-4eb5-80ca-4a89935e6b6c.png)

**Comment by Ein-Tim** [comment]
I just created an EU digital COVID certificate QR-code certificate which has the design of the QR code I shared [here](https://github.com/corona-warn-app/cwa-app-ios/issues/3515#issuecomment-917656837) (no lines, only dots) and CovPassCheck scanned it without problems here...

**Comment by thomasaugsten** [comment]
@merlinseeling can you please use a iOS Screenshot and Display it on a Pcr Monitor to check if the QR code. Because the thicknesses no influence.

**Comment by merlinseeling** [comment]
@Ein-Tim and @thomasaugsten 

Yeah, but no. QR-Codes can be displayed in many ways, but all rely on some simple rules. The thickness makes a difference because the scanning device has to differentiate the different pixels, no matter if they are squares or dots. 

<img width="728" alt="Bildschirmfoto 2021-09-12 um 17 31 58" src="https://user-images.githubusercontent.com/76479756/132993551-6d6b4e35-e58d-496b-9c29-e8231dd086ca.png">

Quelle: "https://en.wikipedia.org/wiki/QR_code"


This QR-code is modified, but only in the thickness. Try scanning it. 

<img width="1119" alt="Bildschirmfoto 2021-09-12 um 18 01 37" src="https://user-images.githubusercontent.com/76479756/132994590-3162afa6-758c-459b-b77a-80498c12f0cf.png">

If you don't want to believe me or just don't want to change it, okay, I would like to find out why it isn't displayed incorrect in the CWA because I see the mistake there. If you see it differently, there is no point in discussing it further here.

**Comment by thomasaugsten** [comment]
Please try to scan a iOS screenshot from a of monitor display. If you don't want to record a video please provide a screenshot to my mail. thomas.augsten@sap.com

**Comment by merlinseeling** [comment]
If I understand you correctly, I should mail you a screenshot of the CWA showing my certificate.

**Comment by thomasaugsten** [comment]
This is correct. On voluntary base of course

**Comment by Ein-Tim** [comment]
PR: #3517 improves this.

**Comment by Ein-Tim** [comment]
#3517 merged into `release/2.9.x`, so I think there will be a hotfix for this.

**Comment by ndegendogo** [comment]
yes, they have already tagged it as release candidate for their testing 🎉🎉🎉

**Comment by Ein-Tim** [comment]
Version 2.9.1 has just been released on GitHub: https://github.com/corona-warn-app/cwa-app-ios/releases/tag/v2.9.1

**Comment by thomasaugsten** [comment]
@merlinseeling Can you please install the update an retest

**Comment by merlinseeling** [comment]
Works properly now, thanks

Should I close the issue?

**Comment by Ein-Tim** [comment]
@merlinseeling 

Yes, it this has been fixed please close it.

**Comment by ndegendogo** [comment]
@merlinseeling thanks for your detailed initial analysis, and for your cooperation to get this reproduced and fixed so fast.
👍👍😁😁🎉🎉

**Comment by dsarkar** [comment]
Dear @merlinseeling and community! Many thanks for contributing here! Best wishes, DS

---
Corona-Warn-App Open Source Team
